src/Distribution/Nixpkgs/Meta.hs

```diff
@@ -85,6 +92,54 @@ nixpkgsPlatformFromCabal (Platform arch os) = "\"" ++ nixArch ++ "-" ++ nixOs ++
             OSX -> "darwin" -- rendered as osx by default
             _   -> CabalPretty.prettyShow os
 
+-- | Obtain a 'NixpkgsPlatform' from a string representation intended for config
+--   files.
+--
+--   * Every string starting with @lib.platforms.@ or @platforms.@ is
+--     parsed into 'NixpkgsPlatformGroup'.
+--   * All other strings are attempted to be interpreted as a system tuple of
+--     the form @ARCH-KERNEL@ and parsed into 'NixpkgsPlatformSingle'.
+--
+--   If none of these formats match the input 'String', 'Nothing' is returned.
+--   A 'Just' result thus only indicates that the format of the platform is
+--   sound — 'nixpkgsPlatformFromString' does /not/ check if the parsed platform
+--   actually exists.
+--
+--   __Note__: 'nixpkgsPlatformFromString' is /not/ the inverse operation for
+--   'NixpkgsPlatform'\'s 'Pretty' instance. It is not intended for parsing Nix
+--   expressions.
+--
+--   >>> nixpkgsPlatformFromString "x86_64-netbsd"
+--   Just (NixpkgsPlatformSingle (Platform X86_64 NetBSD))
+--   >>> nixpkgsPlatformFromString "platforms.riscv"
+--   Just (NixpkgsPlatformGroup (Identifier "riscv"))
+--   >>> nixpkgsPlatformFromString "garbage"
+--   Nothing
+nixpkgsPlatformFromString :: String -> Maybe NixpkgsPlatform
+nixpkgsPlatformFromString s = platformGroup <|> singlePlatform
+  where platformGroup = do
+          -- also accept "platform." as prefix to save some typing
+          name <- stripPrefix "lib.platforms." s <|> stripPrefix "platforms." s
+          Just $ NixpkgsPlatformGroup (ident # name)
+
+        -- we accept anything as a platform tuple which is two non-empty
+        -- strings separated by a '-'
+        splitPlatformTuple = filter (not . null) . splitOn "-"
+        singlePlatform =
+          case splitPlatformTuple s of
+            [arch, os] -> Just $ NixpkgsPlatformSingle
+              $ Platform (parseArch arch) (parseOS os)
+            _ -> Nothing
+
+        parseOS = classifyOS Permissive
+        parseArch as =
+          case classifyArch Strict as of
+            OtherArch "i686" -> I386
+            OtherArch "js" -> JavaScript
+            OtherArch "powerpc" -> PPC
+            OtherArch "powerpc64" -> PPC64
```

Comment:
> Just out of interest, why doesn't classifyArch handle these cases?

It does if you use `Permissive` as `ClassificationStrictness`. The problem is, however, that this is too aggressive, e. g. it would parse `powerpc64le` to `PPC64` which we absolutely don't want.

In general Cabal's `classifyArch` and `classifyOS` behave quite strangely, I think because it expects well-formed LLVM-ish (?) input by default and the main objective seems to be backwards compatibility. E. g. GHC's configure script does a ton of normalization before ever passing the platform string to Cabal which gives me this impression. The main friction point seems to be that things are just named differently between LLVM-inspired Cabal and autotools-inspired nixpkgs, e. g. Cabal uses `ppc64` wheras in nixpkgs that's `powerpc64`.

> Also, although new arches don't turn up often, is it worth having OtherArch _ -> error "implement me" or something?

Not really, there's a lot of arches etc. which don't have their own proper entry in the sum type, so parsing them as `OtherArch` is correct (see also the test cases). Normalization of arch/OS is something I only want to do if necessary, since ideally we take a similar approach to Cabal and expect our users to not pass anything too crazy (`callCabal2nix` will and the `hackage2nix` config always give us a proper nixpkgs platform and “normal” users of `cabal2nix` are hopefully sensible or just not into embedded development with Haskell where all those edgecases appear…).

Reply:
Maybe this could be a comment above the code?

Reply:
Done